### ISSUE TITLE
kable: sanitize column names (feature request)

### ISSUE THREAD

**Issue by stla** [yihui/knitr#1427]
Hello, 

 Similarly to the options `sanitize.colnames.function` and `sanitize.text.function` of `xtable::print.xtable`, it would be nice to have options `escape.colnames` and `escape.text` in `kable`, to respectively escape the column names and the body of the table. Because we sometimes want to escape in the body but not in the header.

**Comment by cderv** [comment]
Yes, you can format using `kableExtra::cell_spec` or `kableExtra::text_spec`. They have an `escape` argument and apply to Latex or HTML. 
Also, `xtable` use `xtable::sanitize` with default options. You can use this function as to prepare your data the print with `kable(..., escape = FALSE)`. 
Personally, I find it interesting that formatting option lives in another package. kableExtra is pretty powerful. You should give it a try to see if it fits your need. 

@stla Could you provide a [reproductible exampe](https://www.tidyverse.org/help/) of what you do with xtable and to have a use case and something to test with ?

**Comment by stla** [comment]
Hmmm it was one year ago, I don't really remember. I think I had a table with some R² in the column names.  The exponent 2 requires the math mode. But yes, I use `kableExtra`.

**Comment by cderv** [comment]
Ok. So if this feature is already in kableExtra, there is no need to work on including it in knitr, isn't it ?

**Comment by stla** [comment]
I didn't know `kableExtra::cell_spec` and `kableExtra::text_spec` and I don't understand their purpose (the help explains nothing). Does this allow to escape separately in the column names and in the cells ? Of course if it is possible there's no need to include that in knitr.

**Comment by cderv** [comment]
> But yes, I use kableExtra.

Sorry, I misunderstood. I thought you had managed to get what you want with `kableExtra`. 

About `cell_spec` and `text_spec`, [docs](http://haozhu233.github.io/kableExtra/awesome_table_in_html.html#celltext_specification) shows that they can be used in combination with _data manipulation package_ to prepare the table, including escaping. 

If you can create a [reprex](https://www.tidyverse.org/help/) of what you want (with `xtable` for example) and what you tried with `kableExtra`, it will be easier for me to test, and see if knitr can evolve to include this. 
Without, I can't do very much.

**Comment by cderv** [comment]
Here is a summary of the situations followed by a reproducible example :

* The feature request is to allow **knitr** to have more control on what is escaped or not, instead of just one switch `TRUE` or `FALSE`. Same as what **xtable** can do
* There is an easy solution: Using helpers to escape data manually when preparing the data. See below with `xtable::sanitize()` and internal `knitr:::escape_latex()`. 
This is really not easy if you have a lot of data like that.
 
````markdown
---
title: test
output:
  pdf_document: 
    keep_tex: TRUE
---

This is the kind of table to print, using special char in colnames and some raw latex in rownames. 

```{r, results='asis'}
money <- matrix(c("$1,000","$900","$100"), ncol = 3,
dimnames = list("$\\alpha$",
c("Income (US$)","Expenses (US$)",
"Profit (US$)")))
```

xtable allow to not escape rownames for example

```{r, results='asis'}
library(xtable)
print(xtable(money))
print(xtable(money), sanitize.rownames.function = identity)
```

But with **knitr**, only all or nothing. So, this will error as $ is not escaped is colnames

```{r}
library(knitr)
options(knitr.table.format = "latex")
```

```{r, eval = FALSE}
kable(money, escape = FALSE)
```

But we need `escape = FALSE` in `kable` for the rownname. 
We can do by preparing the data so it will work

```{r, results='asis'}
money1 <- money
colnames(money1) <- xtable::sanitize(colnames(money1))
money1 <- apply(money1, 2, xtable::sanitize)

kable(money1, escape = FALSE)
```


```{r, results='asis'}
money2 <- money
colnames(money2) <- knitr:::escape_latex(colnames(money2))
money2 <- apply(money2, 2, knitr:::escape_latex)
kable(money2, escape = FALSE)
```
````

Having a way to configure this could be interesting, even if the workaround is data manipulation when you need to output raw latex.

**Comment by yihui** [comment]
Thanks for the request, @stla — and apologies it sat so long.

`kable()` is simple by design, and this kind of fine-grained escaping is better handled either by preparing the data before passing it to `kable()`, or by a richer table package. Two straightforward options:

1. Pre-escape only the parts you want, then use `escape = FALSE`:

    ```r
    money <- matrix(c("$1,000","$900","$100"), ncol = 3,
      dimnames = list("$\\alpha$", c("Income (US$)","Expenses (US$)","Profit (US$)")))
    # escape column names but leave row names as raw LaTeX
    colnames(money) <- knitr:::escape_latex(colnames(money))
    kable(money, format = "latex", escape = FALSE)
    ```

2. Use **kableExtra** (`cell_spec()` / `text_spec()` have their own `escape` argument), which is built for this kind of control.

Adding a second escape switch (or per-part sanitize functions like `xtable`) would grow `kable()`'s API in a direction I'd rather keep out of core knitr. Closing as won't-fix, but the workarounds above should cover the use case.
